File: codeforces/huawei2026/solution.cpp
#include <iostream>
#include <vector>
#include <map>
#include <sstream>
#include <string>
using namespace std;

struct RequestState {
  int rid;
  int assigned; // remote computer assigned
  int stage;    // 0: need P_PRE, 1: need P_PROC, 2: need P_POST, 3: need D_PRE, 4: need D_PROC, 5:
                // need D_POST
  bool waiting_up;
  bool waiting_down;
  int next_stage_after_up;
  int next_stage_after_down;
  int tokens_produced;
  bool finished;

  RequestState(int id, int K)
      : rid(id), assigned(id % K), stage(0), waiting_up(false), waiting_down(false),
        next_stage_after_up(-1), next_stage_after_down(-1), tokens_produced(0), finished(false) {}
};

int main() {
  ios::sync_with_stdio(false);
  cin.tie(nullptr);

  string line;
  // read first line: K S latency_in_ms bandwidth_gbps bytes_per_token num_layers
  if (!getline(cin, line))
    return 0;
  stringstream ss(line);
  int K;
  double S, latency_in_ms, bandwidth_gbps;
  int bytes_per_token, num_layers;
  ss >> K >> S >> latency_in_ms >> bandwidth_gbps >> bytes_per_token >> num_layers;

  // second line: scoring parameters (ignore)
  if (!getline(cin, line))
    return 0;

  // third line: N
  if (!getline(cin, line))
    return 0;
  int N = stoi(line);

  // read N rows of task-time table (ignore)
  for (int i = 0; i < N; ++i) {
    if (!getline(cin, line))
      break;
  }

  bool local_free = true;
  vector<bool> remote_free(K, true);
  map<int, RequestState> requests; // rid -> RequestState

  auto handle_tdn = [&](const string& server, const string& task_spec) {
    // Free the resource that completed the task
    if (server == "E") {
      local_free = true;
    } else if (server.size() > 1 && server[0] == 'C') {
      // server format: C<number>
      int k = stoi(server.substr(1));
      if (k >= 0 && k < K) {
        remote_free[k] = true;
      }
    }

    stringstream ss(task_spec);
    string first;
    ss >> first;
    if (first == "P") {
      string second;
      ss >> second;
      if (second == "PRE") {
        int remote, rid;
        ss >> remote >> rid;
        auto it = requests.find(rid);
        if (it != requests.end() && !it->second.finished) {
          it->second.waiting_up = true;
          it->second.next_stage_after_up = 1; // after UP, can do P_PROC
        }
      } else if (second == "PROC") {
        // ls le remote rid
        int ls, le, remote, rid;
        ss >> ls >> le >> remote >> rid;
        auto it = requests.find(rid);
        if (it != requests.end() && !it->second.finished) {
          it->second.waiting_down = true;
          it->second.next_stage_after_down = 2; // after DOWN, can do P_POST
        }
      } else if (second == "POST") {
        int remote, rid;
        ss >> remote >> rid;
        auto it = requests.find(rid);
        if (it != requests.end() && !it->second.finished) {
          it->second.stage = 3; // ready for D_PRE
          it->second.waiting_up = it->second.waiting_down = false;
          it->second.next_stage_after_up = it->second.next_stage_after_down = -1;
        }
      }
    } else if (first == "D") {
      string second;
      ss >> second;
      if (second == "PRE") {
        int dummy, m;
        ss >> dummy >> m; // dummy is -1
        vector<int> rids(m);
        for (int i = 0; i < m; ++i)
          ss >> rids[i];
        for (int rid : rids) {
          auto it = requests.find(rid);
          if (it != requests.end() && !it->second.finished) {
            it->second.waiting_up = true;
            it->second.next_stage_after_up = 4; // after UP, can do D_PROC
          }
        }
      } else if (second == "PROC") {
        int remote, m;
        ss >> remote >> m;
        vector<int> rids(m);
        for (int i = 0; i < m; ++i)
          ss >> rids[i];
        for (int rid : rids) {
          auto it = requests.find(rid);
          if (it != requests.end() && !it->second.finished) {
            it->second.waiting_down = true;
            it->second.next_stage_after_down = 5; // after DOWN, can do D_POST
          }
        }
      } else if (second == "POST") {
        int dummy, m;
        ss >> dummy >> m; // dummy is -1
        vector<int> rids(m);
        for (int i = 0; i < m; ++i)
          ss >> rids[i];
        for (int rid : rids) {
          auto it = requests.find(rid);
          if (it != requests.end() && !it->second.finished) {
            it->second.tokens_produced += 1;
            // ready for another D_PRE (assuming more tokens)
            it->second.stage = 3;
            it->second.waiting_up = it->second.waiting_down = false;
            it->second.next_stage_after_up = it->second.next_stage_after_down = -1;
          }
        }
      }
    }
  };

  auto handle_xdn = [&](const string& direction, int remote, int size, const string& typ, int m,
                        const vector<int>& rids) {
    for (int rid : rids) {
      auto it = requests.find(rid);
      if (it == requests.end() || it->second.finished)
        continue;
      if (typ == "PRE") {
        if (direction == "UP") {
          if (it->second.waiting_up) {
            it->second.waiting_up = false;
            it->second.stage = it->second.next_stage_after_up;
            it->second.next_stage_after_up = it->second.next_stage_after_down = -1;
          }
        } else { // DOWN
          if (it->second.waiting_down) {
            it->second.waiting_down = false;
            it->second.stage = it->second.next_stage_after_down;
            it->second.next_stage_after_up = it->second.next_stage_after_down = -1;
          }
        }
      } else { // DEC
        if (direction == "UP") {
          if (it->second.waiting_up) {
            it->second.waiting_up = false;
            it->second.stage = it->second.next_stage_after_up;
            it->second.next_stage_after_up = it->second.next_stage_after_down = -1;
          }
        } else { // DOWN
          if (it->second.waiting_down) {
            it->second.waiting_down = false;
            it->second.stage = it->second.next_stage_after_down;
            it->second.next_stage_after_up = it->second.next_stage_after_down = -1;
          }
        }
      }
    }
  };

  auto handle_fin = [&](int rid) {
    auto it = requests.find(rid);
    if (it != requests.end()) {
      it->second.finished = true;
    }
  };

  // function to try schedule one task, returns true if scheduled and pushes assignment to out
  auto try_schedule = [&](vector<string>& out) -> bool {
    // local computer
    if (local_free) {
      for (auto& kv : requests) {
        RequestState& req = kv.second;
        if (req.finished)
          continue;
        if (req.waiting_up || req.waiting_down)
          continue;
        if (req.stage == 0) { // need P_PRE
          out.push_back("E P PRE " + to_string(req.assigned) + " " + to_string(req.rid));
          local_free = false;
          return true;
        } else if (req.stage == 2) { // need P_POST
          out.push_back("E P POST " + to_string(req.assigned) + " " + to_string(req.rid));
          local_free = false;
          return true;
        } else if (req.stage == 3) { // need D_PRE
          out.push_back("E D PRE -1 1 " + to_string(req.rid));
          local_free = false;
          return true;
        } else if (req.stage == 5) { // need D_POST
          out.push_back("E D POST -1 1 " + to_string(req.rid));
          local_free = false;
          return true;
        }
      }
    }
    // remote computers
    for (int k = 0; k < K; ++k) {
      if (remote_free[k]) {
        for (auto& kv : requests) {
          RequestState& req = kv.second;
          if (req.finished)
            continue;
          if (req.waiting_up || req.waiting_down)
            continue;
          if (req.assigned != k)
            continue;
          if (req.stage == 1) { // need P_PROC
            out.push_back("C" + to_string(k) + " P PROC 0 " + to_string(num_layers) + " " +
                          to_string(k) + " " + to_string(req.rid));
            remote_free[k] = false;
            return true;
          } else if (req.stage == 4) { // need D_PROC
            out.push_back("C" + to_string(k) + " D PROC " + to_string(k) + " 1 " +
                          to_string(req.rid));
            remote_free[k] = false;
            return true;
          }
        }
      }
    }
    return false;
  };

  // main loop
  while (true) {
    string timestamp_line;
    if (!getline(cin, timestamp_line))
      break;
    if (timestamp_line == "END")
      break;

    string event_count_line;
    if (!getline(cin, event_count_line))
      break;
    int e;
    {
      stringstream ss(event_count_line);
      if (!(ss >> e))
        break;
    }

    vector<string> events;
    for (int i = 0; i < e; ++i) {
      string event_line;
      if (!getline(cin, event_line))
        break;
      events.push_back(event_line);
    }
    // process events
    for (const string& ev : events) {
      if (ev.empty())
        continue;
      if (ev.substr(0, 3) == "ARR") {
        stringstream ss(ev);
        string dummy;
        int rid, lin;
        ss >> dummy >> rid >> lin;
        if (requests.find(rid) == requests.end()) {
          requests.emplace(rid, RequestState(rid, K));
        }
      } else if (ev.substr(0, 3) == "TDN") {
        // split into server, task_spec, dur (ignore dur)
        size_t first_space = ev.find(' ');
        size_t second_space = ev.find(' ', first_space + 1);
        if (second_space == string::npos)
          continue;
        string server = ev.substr(0, first_space);
        string task_spec = ev.substr(first_space + 1, second_space - first_space - 1);
        // dur = ev.substr(second_space + 1);
        handle_tdn(server, task_spec);
      } else if (ev.substr(0, 3) == "XDN") {
        stringstream ss(ev);
        string direction;
        int remote, size;
        string typ;
        int m;
        ss >> direction >> remote >> size >> typ >> m;
        vector<int> rids(m);
        for (int i = 0; i < m; ++i)
          ss >> rids[i];
        handle_xdn(direction, remote, size, typ, m, rids);
      } else if (ev.substr(0, 3) == "FIN") {
        stringstream ss(ev);
        string dummy;
        int rid;
        ss >> dummy >> rid;
        handle_fin(rid);
      }
      // else ignore
    }
    // after processing events, try to schedule as many tasks as possible in this frame
    vector<string> assignments;
    while (try_schedule(assignments)) {
      // continue scheduling until no more can be scheduled
    }
    // output count and assignments
    cout << assignments.size() << '\n';
    for (const string& assign : assignments) {
      cout << assign << '\n';
    }
    cout << flush;
  }

  return 0;
}

File: codeforces/huawei2026/solution.py
import sys

def solve():
    data = sys.stdin.readline().strip()
    if not data:
        return
    # first line: K S latency_in_ms bandwidth_gbps bytes_per_token num_layers
    parts = data.split()
    K = int(parts[0])
    S = float(parts[1])  # schedule cost, not used in baseline
    latency_in_ms = float(parts[2])
    bandwidth_gbps = float(parts[3])
    bytes_per_token = int(parts[4])
    num_layers = int(parts[5])

    # second line: scoring parameters (ignore)
    sys.stdin.readline()
    # third line: N
    line = sys.stdin.readline().strip()
    if not line:
        return
    N = int(line)
    # read N rows of task-time table (ignore for baseline)
    for _ in range(N):
        sys.stdin.readline()

    # state
    local_free = True
    remote_free = [True] * K

    class RequestState:
        __slots__ = ('rid', 'assigned', 'stage',
                     'waiting_up', 'waiting_down',
                     'next_stage_after_up', 'next_stage_after_down',
                     'tokens_produced', 'finished')
        def __init__(self, rid):
            self.rid = rid
            self.assigned = rid % K  # simple round-robin assignment
            self.stage = 0  # 0: need P_PRE, 1: need P_PROC, 2: need P_POST, 3: need D_PRE, 4: need D_PROC, 5: need D_POST
            self.waiting_up = False
            self.waiting_down = False
            self.next_stage_after_up = None
            self.next_stage_after_down = None
            self.tokens_produced = 0
            self.finished = False

    requests = {}  # rid -> RequestState

    def schedule_p_pre(req):
        nonlocal local_free
        if not local_free:
            return False
        # task: P PRE <remote> <rid>
        print(f"E P PRE {req.assigned} {req.rid}")
        local_free = False
        return True

    def schedule_p_proc(req):
        nonlocal remote_free
        if not remote_free[req.assigned]:
            return False
        # task: P PROC 0 <num_layers> <remote> <rid>
        print(f"C{req.assigned} P PROC 0 {num_layers} {req.assigned} {req.rid}")
        remote_free[req.assigned] = False
        return True

    def schedule_p_post(req):
        nonlocal local_free
        if not local_free:
            return False
        # task: P POST <remote> <rid>
        print(f"E P POST {req.assigned} {req.rid}")
        local_free = False
        return True

    def schedule_d_pre(req):
        nonlocal local_free
        if not local_free:
            return False
        # task: D PRE -1 1 <rid>
        print(f"E D PRE -1 1 {req.rid}")
        local_free = False
        return True

    def schedule_d_proc(req):
        nonlocal remote_free
        if not remote_free[req.assigned]:
            return False
        # task: D PROC <remote> 1 <rid>
        print(f"C{req.assigned} D PROC {req.assigned} 1 {req.rid}")
        remote_free[req.assigned] = False
        return True

    def schedule_d_post(req):
        nonlocal local_free
        if not local_free:
            return False
        # task: D POST -1 1 <rid>
        print(f"E D POST -1 1 {req.rid}")
        local_free = False
        return True

    def handle_tdn(server, task_spec):
        parts = task_spec.split()
        if parts[0] == 'P':
            if parts[1] == 'PRE':
                remote = int(parts[2])
                rid = int(parts[3])
                req = requests.get(rid)
                if req and not req.finished:
                    # after P PRE TDN: queued UP transfer
                    req.waiting_up = True
                    req.next_stage_after_up = 1  # after UP, can do P_PROC
            elif parts[1] == 'PROC':
                # ls le remote rid
                rid = int(parts[5])
                req = requests.get(rid)
                if req and not req.finished:
                    # after P PROC TDN: queued DOWN transfer
                    req.waiting_down = True
                    req.next_stage_after_down = 2  # after DOWN, can do P_POST
            elif parts[1] == 'POST':
                remote = int(parts[2])
                rid = int(parts[3])
                req = requests.get(rid)
                if req and not req.finished:
                    # after P POST TDN: ready for D_PRE
                    req.stage = 3
                    req.waiting_up = False
                    req.waiting_down = False
                    req.next_stage_after_up = None
                    req.next_stage_after_down = None
        elif parts[0] == 'D':
            if parts[1] == 'PRE':
                # D PRE -1 m rid...
                m = int(parts[3])
                rids = list(map(int, parts[4:4+m]))
                for rid in rids:
                    req = requests.get(rid)
                    if req and not req.finished:
                        # after D PRE TDN: queued UP transfer(s)
                        req.waiting_up = True
                        req.next_stage_after_up = 4  # after UP, can do D_PROC
            elif parts[1] == 'PROC':
                remote = int(parts[2])
                m = int(parts[3])
                rids = list(map(int, parts[4:4+m]))
                for rid in rids:
                    req = requests.get(rid)
                    if req and not req.finished:
                        # after D PROC TDN: queued DOWN transfer
                        req.waiting_down = True
                        req.next_stage_after_down = 5  # after DOWN, can do D_POST
            elif parts[1] == 'POST':
                m = int(parts[3])
                rids = list(map(int, parts[4:4+m]))
                for rid in rids:
                    req = requests.get(rid)
                    if req and not req.finished:
                        # after D POST TDN: token produced
                        req.tokens_produced += 1
                        # ready for another D_PRE (assuming more tokens)
                        req.stage = 3
                        req.waiting_up = False
                        req.waiting_down = False
                        req.next_stage_after_up = None
                        req.next_stage_after_down = None
                        # Note: we do not know if more tokens; we will rely on FIN to stop

    def handle_xdn(direction, remote, size, typ, m, rids):
        # direction: 'UP' or 'DOWN'
        # typ: 'PRE' or 'DEC'
        for rid in rids:
            req = requests.get(rid)
            if not req or req.finished:
                continue
            if typ == 'PRE':
                if direction == 'UP':
                    # input-stage UP transfer completed
                    if req.waiting_up:
                        req.waiting_up = False
                        req.stage = req.next_stage_after_up
                        req.next_stage_after_up = None
                else:  # DOWN
                    if req.waiting_down:
                        req.waiting_down = False
                        req.stage = req.next_stage_after_down
                        req.next_stage_after_down = None
            else:  # DEC
                if direction == 'UP':
                    # output-step UP transfer completed
                    if req.waiting_up:
                        req.waiting_up = False
                        req.stage = req.next_stage_after_up
                        req.next_stage_after_up = None
                else:  # DOWN
                    if req.waiting_down:
                        req.waiting_down = False
                        req.stage = req.next_stage_after_down
                        req.next_stage_after_down = None

    def handle_fin(rid):
        req = requests.get(rid)
        if req:
            req.finished = True
            # optionally clean up

    def try_schedule():
        nonlocal local_free, remote_free
        scheduled_any = False
        # local computer
        if local_free:
            for rid, req in list(requests.items()):
                if req.finished:
                    continue
                if req.waiting_up or req.waiting_down:
                    continue
                if req.stage == 0:  # need P_PRE
                    if schedule_p_pre(req):
                        scheduled_any = True
                        break
                elif req.stage == 2:  # need P_POST
                    if schedule_p_post(req):
                        scheduled_any = True
                        break
                elif req.stage == 3:  # need D_PRE
                    if schedule_d_pre(req):
                        scheduled_any = True
                        break
                elif req.stage == 5:  # need D_POST
                    if schedule_d_post(req):
                        scheduled_any = True
                        break
        # remote computers
        for k in range(K):
            if remote_free[k]:
                for rid, req in list(requests.items()):
                    if req.finished:
                        continue
                    if req.waiting_up or req.waiting_down:
                        continue
                    if req.assigned != k:
                        continue
                    if req.stage == 1:  # need P_PROC
                        if schedule_p_proc(req):
                            scheduled_any = True
                            break
                    elif req.stage == 4:  # need D_PROC
                        if schedule_d_proc(req):
                            scheduled_any = True
                            break
        return scheduled_any

    # main loop
    while True:
        line = sys.stdin.readline()
        if not line:
            break
        line = line.strip()
        if line == 'END':
            break
        # timestamp line (ignore)
        t = line
        # event count line
        line = sys.stdin.readline()
        if not line:
            break
        e = int(line.strip())
        events = []
        for _ in range(e):
            events.append(sys.stdin.readline().rstrip('\n'))
        # process events
        for ev in events:
            if not ev:
                continue
            if ev.startswith('ARR'):
                # ARR rid L_in
                _, rid_str, lin_str = ev.split()
                rid = int(rid_str)
                # L_in not needed
                if rid not in requests:
                    requests[rid] = RequestState(rid)
                else:
                    # should not happen
                    pass
            elif ev.startswith('TDN'):
                # TDN server task_spec dur
                # split into three parts: server, task_spec, dur
                # but task_spec may contain spaces, so split max 2
                parts = ev.split(' ', 2)
                if len(parts) < 3:
                    continue
                server = parts[0]
                task_spec = parts[1]
                # dur = parts[2]  # ignore
                handle_tdn(server, task_spec)
            elif ev.startswith('XDN'):
                # XDN UP/DOWN remote size PRE/DEC m rid...
                parts = ev.split()
                if len(parts) < 6:
                    continue
                direction = parts[0]
                remote = int(parts[1])
                size = int(parts[2])  # bytes, not used
                typ = parts[3]
                m = int(parts[4])
                rids = list(map(int, parts[5:5+m]))
                handle_xdn(direction, remote, size, typ, m, rids)
            elif ev.startswith('FIN'):
                # FIN rid
                _, rid_str = ev.split()
                rid = int(rid_str)
                handle_fin(rid)
            else:
                # unknown, ignore
                pass
        # after processing events, try to schedule tasks
        # we may be able to schedule multiple tasks in same frame (different resources)
        # keep trying until no more can be scheduled
        while True:
            if not try_schedule():
                break
        sys.stdout.flush()

if __name__ == "__main__":
    solve()